### ISSUE TITLE
Epic: adopt the Bunsen standard library (burn 0.21) — machine-checked shape contracts on the FEM tensor pipeline + opportunistic ops

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/geode-fem#355]
# Epic: Adopt Bunsen (the burn standard library)

## Context

[Bunsen](https://zspacelabs.ai/bunsen/book/) is the community **standard library for the [`burn`](https://burn.dev) tensor framework** — the same framework GEODE-FEM is built on. It bundles, into one versioned dependency, components that extend burn's core: runtime shape **contracts**, extra tensor **ops**, NN **blocks**, domain **kits**, and burn-adjacent **burner** infra.

**The load-bearing feasibility fact:** Bunsen's version tracks burn's (`bunsen 0.21` ⇄ `burn 0.21`), and **GEODE-FEM already pins `burn = 0.21.0`** (`Cargo.toml:17`, confirmed in `Cargo.lock`). So adoption needs **no burn bump** — the highest-risk axis for a dependency like this is already green. Bunsen is not yet a dependency.

## Why this is worth doing (and where it is NOT)

I evaluated all five Bunsen layers against what GEODE-FEM actually is — a **deterministic FEM/DG electromagnetics solver**, not a trained ML model. The fit is uneven, so this epic is deliberately scoped to the parts that pay off and explicitly rules out the parts that don't.

### In scope — real value

**1. `bunsen::contracts` — machine-checked tensor-shape validation (PRIMARY GOAL).**
The codebase has **~139 ad-hoc `.dims()` / shape-assert sites** in `crates/geode-core/src`, concentrated in the 6 Tensor-using files: `assembly.rs`, `nedelec_assembly.rs`, `nedelec.rs`, `p1.rs`, `eigen.rs`, `lib.rs`. These are exactly the FEM assembly → element-matrix → global-operator → eigensolve pipeline, where a transposed or mis-sized tensor is a real and recurring bug class. Bunsen's `shape_contract!` DSL turns paper-style shape notation into `static`, machine-checked constraints:
```rust
static M_LOCAL: ShapeContract = shape_contract!["n_dof", "n_dof"];
let [ndof, _] = unpack_shape_contract!(M_LOCAL, &m.shape(), &["n_dof"], &[]);
```
- `no_std`-compatible; ~160 ns/unpack on a 4D shape; patterns compile to `static` (zero per-call construction).
- `assert_shape_contract_periodically!` (exponential backoff, ~4 ns amortized) for hot assembly loops.
- With the `burner` feature the `ShapeView` trait validates `&Tensor<B, R, K>` directly.
- Error messages are diagnostic (`8 !~ height=(h_wins*window) :: No integer solution`), far better than a bare `assert_eq!(a.dims()[0], b.dims()[1])`.

**Payoff:** the FEM math contracts (e.g. "the local mass matrix is `n_dof × n_dof`", "the element coordinate tensor is `n_nodes × 3`") become self-documenting, machine-checked, and KaTeX-documentable — replacing scattered, terse asserts with one readable source of truth.

**2. `bunsen::ops` — opportunistic, evaluate-only (SECONDARY).**
`float_arange` / `float_linspace` / `ClampOp` / `repeat_interleave`. Note: the core currently has **0** `arange`/`linspace` call sites, so this is **not** a dedup win today — it's a "reach for these instead of hand-rolling" convention for *future* frequency-sweep / coordinate-generation code. Low priority; adopt where it naturally lands, don't refactor to force it.

### Explicitly OUT of scope (documented to prevent overreach)

- **`bunsen::blocks`** (attention, rotary embeddings, pooling, dropout regularization) — pure neural-network modules. GEODE-FEM has **no** NN model. Not applicable.
- **`bunsen::burner` composite optimizers** (AdamW/Muon `OptimizerGroup`, `Learner` integration) — GEODE-FEM has **0** `Module`/`Optimizer`/`burn::train` usage; it's a deterministic physics solver with no training loop. Not applicable. (The `burner` *feature flag* is still pulled in solely to let `contracts` validate `&Tensor`, which is a different, lightweight use.)
- **`bunsen::kits::sims`** (Conway's Game of Life, Lattice-Boltzmann D2Q9) — these are specific cellular/fluid kernels, **not** a generic iteration framework ("the caller manages iteration loops; no fixed-point/convergence abstractions"). They do **not** model EM FEM and do **not** replace our own `iterate_while` combinator (#301). Not applicable.

## Proposed phases (child issues to spin off after approval)

- **Phase 1 — Spike + dependency wiring.** Add `bunsen = "0.21"` with the `contracts` + `burner` features to the workspace; confirm it resolves against the existing `burn 0.21.0` lock with **no** transitive bump and no feature conflict (we run `--no-default-features` burn with CPU backends — verify Bunsen doesn't drag in wgpu/cuda or a `std`/`no_std` conflict). Convert **one** representative shape-assert cluster in `assembly.rs` to `shape_contract!` as a proof of value. Gate: builds clean on the arpack/ndarray CI feature combos; `cargo doc -D warnings` (doc-lint gate) stays green.
- **Phase 2 — Contracts across the assembly path.** Systematically replace the ad-hoc `.dims()` asserts in `assembly.rs` + `nedelec_assembly.rs` with named `shape_contract!`s carrying the FEM math notation; use `assert_shape_contract_periodically!` in the per-element hot loops. Document each contract with the KaTeX it encodes.
- **Phase 3 — Contracts on basis + eigensolver.** Extend to `nedelec.rs`, `p1.rs`, `eigen.rs` (element basis tensors, the dense/sparse pencil shapes feeding `FaerDenseEigensolver` / the ARPACK path).
- **Phase 4 — Ops convention (optional/evaluate).** Adopt `float_linspace`/`ClampOp` where new range/clamp code appears (e.g. future dispersion frequency sweeps); add a short CONTRIBUTING note pointing future tensor code at `bunsen::ops` first. May be closed as "convention documented, no refactor needed."

## Risks / open questions
- **Feature-flag surface:** GEODE-FEM builds burn with `default-features = false, features = ["std"]` and CPU backends (ndarray/arpack). Phase 1 must prove Bunsen's `contracts`+`burner` features don't pull a heavier backend or break the `std`-only build. This is the one real integration risk.
- **Lockstep upgrades:** adopting Bunsen couples us to the `burn`↔`bunsen` 0.x release cadence — a future burn bump now also needs a matching bunsen bump. Acceptable (we're already on burn's cadence) but worth stating.
- **Value concentration:** only 6 files use `Tensor<`; the contracts win is real but bounded to the assembly/eigensolve core, not the whole crate (much of the FEM core is sparse/faer/ndarray, outside Bunsen's tensor surface).

## Epic acceptance criteria
- [ ] Bunsen wired in at `0.21` against the existing `burn 0.21.0` with no transitive/backend regressions; CI (arpack, ndarray, doc-lint, rustfmt) stays green.
- [ ] The ad-hoc shape asserts in the assembly path (`assembly.rs`, `nedelec_assembly.rs`) are migrated to `shape_contract!` with documented math notation; hot loops use the periodic variant.
- [ ] Basis + eigensolver shape checks (`nedelec.rs`, `p1.rs`, `eigen.rs`) migrated.
- [ ] `bunsen::ops` either adopted where natural or explicitly documented as a forward convention.
- [ ] Out-of-scope layers (`blocks`, `burner` optimizers, `kits::sims`) remain unused, with the rationale recorded.

---
*Filed at operator request after reviewing the Bunsen book (https://zspacelabs.ai/bunsen/book/). Architect/Champion: please validate the phase breakdown and the burn-0.21 exact-match assumption before spinning off Phase 1.*

**Comment by crutcher** [comment]
When using shape contracts(as used in the bunsen blocks and kits); we *could* use static contract defs, but there are mechanisms which reduce the code usage.

`assert_shape_contract!(...);` provides a way to define an anonymous contract; and then immediately use it.
`periodically_assert_shape_contract!(...);` provides a way to do the same, but trigger less often.

`let [a, b] = unpack_shape_contract!([...], &thing, &["a", "b"], &[(.., ..), ...]);` provides a mechanism to resolve a shape expression, and extract a specific subset.

Truly perf-critical blocks can use `#[cfg(debug_assertions)]` for all contracts in code.

**Comment by rjwalters** [comment]
## Architect validation pass (pre-Phase-1)

Validated the load-bearing feasibility claims against the registry index + the live tree. Summary: **the central "no burn bump" thesis HOLDS, but three integration details in the body are wrong and would have broken Phase 1 as written.**

**Held:**
- `burn 0.21.0` pinned — `Cargo.toml:17` + `Cargo.lock`. ✅
- `bunsen 0.21.0` exists, requires `burn ^0.21.0` (semver-compatible with our pin, no bump). ✅ Latest is 0.24.2, also on `burn ^0.21.0`.
- The 6-file Tensor set (`assembly.rs`, `nedelec_assembly.rs`, `nedelec.rs`, `p1.rs`, `eigen.rs`, `lib.rs`) is correct. ✅

**Corrected:**
- ❌ **No `contracts` / `burner` features exist on `bunsen 0.21.x`.** The proposed `features = ["contracts", "burner"]` will NOT resolve. Contracts ship via the always-on `bunsen-contracts-macros` dep, gated only by `std`. Correct wiring: `default-features = false, features = ["std"]`.
- ❌ **bunsen's DEFAULT features are `["reflection", "train", "testing"]`** — `train` pulls `burn/train` (explicitly out-of-scope here) and `reflection` pulls `xot` + `xee-xpath` (an XPath engine). Must use `default-features = false`.
- ⚠️ **"build burn with default-features=false, CPU backends" is half-right** — workspace sets that, but `geode-core`'s own `default = ["wgpu"]`; CI runs `--no-default-features --features arpack,ndarray`. Phase 1 must verify both combos.
- ⚠️ **~139 shape-assert sites is high** — measured 31 `.dims()` + 64 `assert*` = 95 in the 6 files (0 `.shape()`).
- ⚠️ **Prefer `bunsen 0.21.x` over `0.24.x`** — 0.24 adds a *mandatory* `downloader` dep even with `default-features = false`; 0.21.x does not.

**Phase breakdown:** sound as-is. Phase 1 is correctly the gate (the no-backend-drag check decides whether 2–4 proceed). Filed Phase 1 only.

➡️ **Child issue: #358** (Bunsen Phase 1 — dependency-wiring spike + one `assembly.rs` shape_contract!). Acceptance criteria there reflect the corrected wiring.

**Comment by crutcher** [comment]
Bunsen 0.25.0 has been pushed.

**Comment by rjwalters** [comment]
## Decomposition: Phases 2–4 filed

Filed the remaining child implementation issues for the phase breakdown. All are `loom:triage` + `tier:goal-supporting` + `loom:blocked`.

- **#453** — Phase 2: shape contracts across the assembly path (`crates/geode-core/src/assembly/p1.rs`, `crates/geode-core/src/assembly/nedelec.rs`)
- **#454** — Phase 3: shape contracts on basis + eigensolver (`crates/geode-core/src/elements/p1.rs`, `crates/geode-core/src/elements/nedelec.rs`, `crates/geode-core/src/eigen/dense.rs`)
- **#455** — Phase 4: `bunsen::ops` convention (evaluate-only; may close as "convention documented, no refactor needed")

### Dependency chain

```
#358 (Phase 1 spike — BLOCKED upstream)
  └─▶ #453 (Phase 2)
        └─▶ #454 (Phase 3)
        └─▶ #455 (Phase 4)
```

All three are gated on **#358** (Phase 1 dependency-wiring spike), which establishes the actual `bunsen` wiring in `geode-core/Cargo.toml`. #454 and #455 additionally depend on **#453 (Phase 2)** merging first — #454 for the contract-naming/documentation conventions Phase 2 establishes, #455 because `bunsen` isn't a `geode-core` dependency at all until Phase 2 lands it.

### Standing upstream gate (unchanged from #358)

#358 is `loom:blocked`: bunsen 0.21.x has an ungated `pub use hashbrown;` (upstream bug zspacelabs/bunsen#95, closed 2026-06-24) that makes any non-`train` build fail to compile — so the epic's required `default-features = false, features = ["std"]` wiring cannot compile on 0.21.x. The fix shipped in **bunsen 0.25.0**, but 0.25.0 adds a mandatory (non-optional) `downloader` + `directories-next` dependency that fails the no-drag gate as currently specified. No fixed 0.21.x release exists. See #358's 2026-07-12 blocker re-check comment for the three unblock options (ask upstream to gate `downloader` behind a feature; ask for a 0.21.4 hashbrown-fix backport; or relax the no-drag gate to tolerate `downloader`).

None of #453/#454/#455 can start until that gate clears (or the epic/Champion picks one of the unblock paths).

### File-layout correction

The epic body and #358 both predate refactors #389/#390/#397/#413, which split the flat `assembly.rs` / `nedelec_assembly.rs` / `nedelec.rs` / `p1.rs` / `eigen.rs` files into `assembly/{p1,nedelec,fe,sparse,surface}.rs`, `elements/{p1,nedelec}.rs`, and `eigen/{mod,dense,arpack,lanczos,self_consistent}.rs` (+ `eigen/complex/*`). #453 and #454 target the current file locations and were re-measured against the live tree (2026-07-13):

- Phase 2 surface: `assembly/p1.rs` (4 `.dims()`) + `assembly/nedelec.rs` (13 `.dims()` + 40 `assert*`) = **57 sites**
- Phase 3 surface: `elements/p1.rs` (1+2) + `elements/nedelec.rs` (12+13) + `eigen/dense.rs`'s one `Tensor` site (1 `.dims()`) = **35 sites** on the Burn-`Tensor` surface (`eigen/dense.rs`'s remaining 6 asserts are on `faer::Mat`, outside Bunsen's tensor surface — #454 asks for an explicit keep-or-convert decision rather than a forced fit)

Total ≈ 92 sites across both phases, consistent with #358's corrected ~95 estimate (the small delta is the file split, not new code).

**Comment by rjwalters** [comment]
**Phase 4 outcome — `bunsen::ops` convention (evaluate-only), via #455**

Re-verified against `origin/main` @ `ebd57c6` (post-#358/#453 merge) in the issue-455 worktree:

- **`arange`/`linspace` call sites in `crates/geode-core/src`: 0** — `grep -rn -E '\b(arange|linspace)\b'` returns nothing. No retroactive refactor target, consistent with the epic's original claim.
- **Tensor-level clamp candidates: 0.** A `grep -rn '\.clamp' crates/geode-core/src` finds **8 hits**, but every one is a **scalar** `f64`/`f32` clamp inside a per-element loop body — PML ramp fractions (`mesh/patch.rs`, `driven/scattering.rs` ×2, `assembly/nedelec.rs` ×2, `analytic/waveguide.rs` ×2) and a directivity ratio (`postproc/ntff.rs`). `ClampOp::clamp()` wraps `Tensor<B, D>` only, so **none** of the 8 are refactor candidates. The grep count is *not* "8 candidates" — it's still 0 relevant sites.

**Resolution:** no code refactor warranted → **"convention documented, no refactor needed"** per this epic's Phase 4 escape hatch. Since no `CONTRIBUTING.md` exists anywhere in the tree, the convention note landed in the top-level `README.md` (new "Bunsen conventions" subsection under Build): future tensor-index/sequence code should prefer `bunsen::ops` (`float_arange`, `float_linspace`, tensor `ClampOp`, `repeat_interleave`) over hand-rolled Burn primitives; `bunsen::contracts::*` shape-contract macros (template in `assembly/p1.rs`) documented alongside; scalar (non-`Tensor`) math explicitly out of scope.

Documented in PR #468 (`Closes #455`).

**Comment by rjwalters** [comment]
Epic complete — closing on the merge of the final phase child #454 (PR #469).

**All four phases delivered and merged:**
- Phase 1 #358 / PR #466 — dependency-wiring spike: bunsen git-pinned to zspacelabs/bunsen@1ff74cd (0.25.0, `default-features = false, features = ["std"]`), no-drag gate judge-verified (burn stays 0.21.0, zero forbidden edges: no downloader/burn-train/wgpu/xot/xee-xpath compiled), first `shape_contract!` conversion as template. Unblocked by our upstream PR zspacelabs/bunsen#112 (optional downloader) + #95 (hashbrown gate).
- Phase 2 #453 / PR #467 — assembly path: all 15 tensor-shape sites in assembly/p1.rs + assembly/nedelec.rs onto 5 named static contracts with FEM math-notation docs; `assert_shape_contract_periodically!` in per-element hot loops; 40 host-side (Vec::len/scalar) asserts correctly left plain; firing tests prove the contracts trip.
- Phase 3 #454 / PR #469 — basis + eigensolver: 14 tensor sites in elements/p1.rs, elements/nedelec.rs, eigen/dense.rs onto 6 named contracts; basis kernels verified COLD (batched, once per assembly) so full-check variant is correct; faer::Mat asserts correctly out of scope; 4 firing tests.
- Phase 4 #455 / PR #468 — bunsen::ops convention: 0 refactor sites (verified: no arange/linspace; the 8 `.clamp` hits are scalar, ClampOp is Tensor-only), convention documented in README per the epic's escape hatch.

**Standing follow-up (not an epic blocker):** swap the git pin to `bunsen = { version = "0.25.1", ... }` when release-plz cuts the release upstream — recorded in PR #466.

**Future-phase candidates (unfiled):** contracts for the new Epic #448 assembly modules (magnetostatic.rs: 6 asserts; torque.rs: 9 asserts) — flagged during #453 curation.

Closed by Loom sweep orchestration.
